File: ts/src/similarity.ts
import { createHash } from "node:crypto";
import { CollisionError, ContainmentError } from "./errors.js";
import type { Node } from "./node.js";
import { RESERVED_NAMESPACE, readJson, writeJsonAtomic } from "./paths.js";
import { scoreKey } from "./ranking.js";

export type Vector = number[];

/** The seam: turns text into vectors. The kernel ships no concrete embedder. */
export interface Embedder {
  readonly cacheNamespace: string;
  embed(texts: string[]): Vector[];
}

/** The frozen per-node embedding input: title and body joined by one blank line. */
export function embedText(node: Node): string {
  return `${node.title}\n\n${node.body}`;
}

/** Content-address key for the vector cache. */
export function textHash(text: string): string {
  return createHash("sha256").update(text, "utf-8").digest("hex");
}

const NAMESPACE_RE = /^[A-Za-z0-9._-]+$/;

/** A cacheNamespace must be a safe single path segment. */
export function validateNamespace(namespace: string): void {
  if (namespace === "." || namespace === ".." || !NAMESPACE_RE.test(namespace)) {
    throw new TypeError(`invalid cacheNamespace ${JSON.stringify(namespace)}`);
  }
}

const TEXT_HASH_RE = /^[0-9a-f]{64}$/;

/** A cache key must be exactly 64 lowercase hex chars (a SHA-256 hexdigest). */
export function validateTextHash(hash: string): void {
  if (!TEXT_HASH_RE.test(hash)) {
    throw new TypeError(`invalid textHash ${JSON.stringify(hash)}`);
  }
}

/** Reject empty / non-numeric / non-finite vectors. Booleans are non-`number` typed,
 * so `typeof x !== "number"` already rejects them. */
function validateFinite(vec: ReadonlyArray<unknown>): void {
  if (vec.length < 1) {
    throw new RangeError("vector must have length >= 1");
  }
  for (const x of vec) {
    if (typeof x !== "number" || !Number.isFinite(x)) {
      throw new TypeError(`vector contains non-finite or non-numeric value ${String(x)}`);
    }
  }
}

/** Return the L2-normalized vector; reject zero-norm and invalid numeric input. */
function normalize(vec: Vector): Vector {
  validateFinite(vec);
  let sumSq = 0;
  for (const x of vec) sumSq += x * x;
  const norm = Math.sqrt(sumSq);
  if (norm === 0) {
    throw new RangeError("cannot normalize a zero-norm vector");
  }
  return vec.map((x) => x / norm);
}

export interface SimilarHit {
  id: string;
  uid: string;
  score: number;
}

export interface VectorSnapshot {
  namespace: string | null;
  dim: number | null;
  vectors: Record<string, number[]>;
  idByUid: Record<string, string>;
  hashByUid: Record<string, string>;
}

function validateK(k?: number): void {
  if (k !== undefined && (!Number.isInteger(k) || k <= 0)) {
    throw new RangeError(`k must be a positive integer or undefined, got ${JSON.stringify(k)}`);
  }
}

/** Content-addressed on-disk cache of RAW embedder output, namespaced per embedder.
 * Disposable: deleting the directory just forces re-embedding. All ranking math lives in
 * VectorIndex; this is purely a model-output cache. */
export class VectorCache {
  readonly root: string;

  constructor(root: string) {
    this.root = root;
  }

  private relFor(namespace: string, hash: string): string {
    validateNamespace(namespace);
    validateTextHash(hash);
    return `${RESERVED_NAMESPACE}/vectors/${namespace}/${hash}.json`;
  }

  get(namespace: string, hash: string): Vector | null {
    const rel = this.relFor(namespace, hash);
    let data: unknown;
    try {
      data = readJson(this.root, rel);
    } catch (e) {
      if (e instanceof ContainmentError) throw e;
      throw new TypeError(`corrupt cache file ${rel}: ${(e as Error).message}`);
    }
    if (data === null) return null;
    if (typeof data !== "object" || data === null || !("dim" in data) || !("vector" in data)) {
      throw new TypeError(`corrupt cache file ${rel}: missing dim/vector`);
    }
    const { dim, vector } = data as { dim: unknown; vector: unknown };
    if (!Array.isArray(vector) || vector.length !== dim) {
      throw new TypeError(`corrupt cache file ${rel}: dim/vector length mismatch`);
    }
    validateFinite(vector);
    return [...(vector as number[])];
  }

  put(namespace: string, hash: string, vector: Vector): void {
    validateFinite(vector);
    writeJsonAtomic(this.root, this.relFor(namespace, hash), { dim: vector.length, vector });
  }
}

interface PreparedVector {
  readonly textHash: string;
  readonly namespace: string;
  readonly vector: Vector | null; // null => content unchanged (id-only refresh)
}

/** In-memory uid -> L2-normalized vector store with exact cosine ranking. Bound to exactly
 * one embedder namespace and one dimension (cosine across vectors from different models or
 * dimensions is meaningless). */
export class VectorIndex {
  vectors = new Map<string, Vector>();
  idByUid = new Map<string, string>();
  hashByUid = new Map<string, string>();
  dim: number | null = null;
  namespace: string | null = null;

  toDict(): VectorSnapshot {
    return {
      namespace: this.namespace,
      dim: this.vectors.size > 0 ? this.dim : null,
      vectors: Object.fromEntries([...this.vectors].map(([uid, vec]) => [uid, [...vec]])),
      idByUid: Object.fromEntries(this.idByUid),
      hashByUid: Object.fromEntries(this.hashByUid),
    };
  }

  static fromDict(d: unknown): VectorIndex {
    if (typeof d !== "object" || d === null) throw new Error("vector snapshot: document must be an object");
    const raw = d as Record<string, unknown>;
    for (const key of ["namespace", "dim", "vectors", "idByUid", "hashByUid"]) {
      if (!(key in raw)) throw new Error(`vector snapshot: missing ${key}`);
    }
    const { namespace, dim } = raw;
    if (typeof raw.vectors !== "object" || raw.vectors === null)
      throw new Error("vector snapshot: vectors must be an object");
    if (typeof raw.idByUid !== "object" || raw.idByUid === null)
      throw new Error("vector snapshot: idByUid must be an object");
    if (typeof raw.hashByUid !== "object" || raw.hashByUid === null)
      throw new Error("vector snapshot: hashByUid must be an object");
    const vectorsRaw = raw.vectors as Record<string, unknown>;
    const idByUid = new Map<string, string>();
    for (const [uid, id] of Object.entries(raw.idByUid as Record<string, unknown>)) {
      if (typeof id !== "string") throw new Error("vector snapshot: idByUid must map string uids to string ids");
      idByUid.set(uid, id);
    }
    const hashByUid = new Map<string, string>();
    for (const [uid, h] of Object.entries(raw.hashByUid as Record<string, unknown>)) {
      if (typeof h !== "string") throw new Error("vector snapshot: hashByUid values must be strings");
      validateTextHash(h);
      hashByUid.set(uid, h);
    }
    const vectorUids = Object.keys(vectorsRaw);
    const sameSet =
      vectorUids.length === idByUid.size &&
      vectorUids.length === hashByUid.size &&
      vectorUids.every((uid) => idByUid.has(uid) && hashByUid.has(uid));
    if (!sameSet) throw new Error("vector snapshot: vectors/idByUid/hashByUid uid sets differ");

    const hasVectors = vectorUids.length > 0;
    if (hasVectors) {
      if (typeof namespace !== "string")
        throw new Error("vector snapshot: namespace must be a string when vectors are present");
      validateNamespace(namespace);
      if (typeof dim !== "number" || !Number.isInteger(dim) || dim <= 0) {
        throw new Error("vector snapshot: dim must be a positive integer when vectors are present");
      }
    } else if (dim !== null) {
      throw new Error("vector snapshot: dim must be null when there are no vectors");
    } else if (namespace !== null) {
      if (typeof namespace !== "string") throw new Error("vector snapshot: namespace must be a string when non-null");
      validateNamespace(namespace);
    }

    const vectors = new Map<string, Vector>();
    for (const [uid, rawVec] of Object.entries(vectorsRaw)) {
      if (!Array.isArray(rawVec)) throw new Error("vector snapshot: vector must be an array");
      for (const x of rawVec) {
        if (typeof x !== "number" || !Number.isFinite(x))
          throw new Error("vector snapshot: vector contains a non-finite value");
      }
      const vec = rawVec as number[];
      if (vec.length !== dim) throw new Error("vector snapshot: vector length != dim");
      const norm = Math.sqrt(vec.reduce((s, x) => s + x * x, 0));
      if (Math.abs(norm - 1) > 1e-9) throw new Error("vector snapshot: stored vector must be L2-normalized");
      vectors.set(uid, [...vec]);
    }

    const idx = new VectorIndex();
    idx.namespace = (namespace as string | null) ?? null;
    idx.vectors = vectors;
    idx.idByUid = idByUid;
    idx.hashByUid = hashByUid;
    idx.dim = (dim as number | null) ?? null;
    return idx;
  }

  static build(nodes: Iterable<Node>, embedder: Embedder, cache: VectorCache): VectorIndex {
    const idx = new VectorIndex();
    validateNamespace(embedder.cacheNamespace);
    idx.namespace = embedder.cacheNamespace; // bind even for an empty corpus
    for (const node of nodes) {
      if (idx.hashByUid.has(node.uid)) {
        throw new CollisionError(`duplicate uid ${JSON.stringify(node.uid)} in corpus`);
      }
      idx.upsert(node, embedder, cache);
    }
    return idx;
  }

  /** Resolve + validate the vector WITHOUT mutating index state (cache writes ok). */
  prepare(node: Node, embedder: Embedder, cache: VectorCache): PreparedVector {
    const namespace = embedder.cacheNamespace;
    validateNamespace(namespace);
    if (this.namespace !== null && namespace !== this.namespace) {
      throw new RangeError(
        `embedder namespace ${JSON.stringify(namespace)} != index namespace ${JSON.stringify(this.namespace)}`,
      );
    }
    const text = embedText(node);
    const h = textHash(text);
    if (this.hashByUid.get(node.uid) === h) {
      return { textHash: h, namespace, vector: null };
    }
    let raw = cache.get(namespace, h);
    if (raw === null) {
      const embedded = embedder.embed([text]);
      if (embedded.length !== 1) {
        throw new RangeError(`embedder returned ${embedded.length} vectors for 1 input`);
      }
      raw = embedded[0];
      validateFinite(raw);
      cache.put(namespace, h, raw);
    }
    if (this.dim !== null && raw.length !== this.dim) {
      throw new RangeError(`vector dim ${raw.length} != index dim ${this.dim}`);
    }
    return { textHash: h, namespace, vector: normalize(raw) };
  }

  /** Apply a prepared vector. Infallible: never throws on valid prepared input. */
  commit(node: Node, prepared: PreparedVector): void {
    if (this.namespace === null) this.namespace = prepared.namespace;
    if (prepared.vector === null) {
      this.idByUid.set(node.uid, node.id); // rename / id-only refresh
      return;
    }
    if (this.dim === null) this.dim = prepared.vector.length;
    this.vectors.set(node.uid, prepared.vector);
    this.idByUid.set(node.uid, node.id);
    this.hashByUid.set(node.uid, prepared.textHash);
  }

  upsert(node: Node, embedder: Embedder, cache: VectorCache): void {
    this.commit(node, this.prepare(node, embedder, cache));
  }

  remove(uid: string): void {
    this.vectors.delete(uid);
    this.idByUid.delete(uid);
    this.hashByUid.delete(uid);
  }

  queryVector(vec: Vector, k?: number): SimilarHit[] {
    validateK(k);
    return this.rank(this.prepareQuery(vec), k, null);
  }

  similar(uid: string, k?: number): SimilarHit[] {
    validateK(k);
    const vec = this.vectors.get(uid);
    if (vec === undefined) throw new Error(`uid ${JSON.stringify(uid)} not in vector index`);
    return this.rank(vec, k, uid);
  }

  similarText(text: string, embedder: Embedder, k?: number): SimilarHit[] {
    validateK(k);
    if (this.namespace !== null && embedder.cacheNamespace !== this.namespace) {
      throw new RangeError(
        `embedder namespace ${JSON.stringify(embedder.cacheNamespace)} != index namespace ${JSON.stringify(this.namespace)}`,
      );
    }
    const embedded = embedder.embed([text]);
    if (embedded.length !== 1) {
      throw new RangeError(`embedder returned ${embedded.length} vectors for 1 input`);
    }
    return this.queryVector(embedded[0], k);
  }

  private prepareQuery(vec: Vector): Vector {
    validateFinite(vec);
    if (this.dim !== null && vec.length !== this.dim) {
      throw new RangeError(`query dim ${vec.length} != index dim ${this.dim}`);
    }
    return normalize(vec);
  }

  private rank(queryVec: Vector, k: number | undefined, excludeUid: string | null): SimilarHit[] {
    const hits: SimilarHit[] = [];
    for (const [uid, vec] of this.vectors) {
      if (uid === excludeUid) continue;
      let dot = 0;
      for (let i = 0; i < queryVec.length; i++) dot += queryVec[i] * vec[i];
      hits.push({ id: this.idByUid.get(uid) as string, uid, score: dot });
    }
    hits.sort((a, b) => {
      const ka = scoreKey(a.score);
      const kb = scoreKey(b.score);
      if (ka !== kb) return kb - ka; // scoreKey descending
      return a.id < b.id ? -1 : a.id > b.id ? 1 : 0; // id ascending
    });
    return k === undefined ? hits : hits.slice(0, k);
  }
}

File: python/src/nodes/core/similarity.py
from __future__ import annotations

import hashlib
import math
import re
from collections.abc import Iterable
from dataclasses import dataclass
from pathlib import Path
from typing import Protocol

from nodes.core.errors import CollisionError
from nodes.core.node import Node
from nodes.core.paths import RESERVED_NAMESPACE, read_json, write_json_atomic
from nodes.core.ranking import score_key

Vector = tuple[float, ...]

_NAMESPACE_RE = re.compile(r"[A-Za-z0-9._-]+")
_VECTOR_SNAPSHOT_KEYS = frozenset({"namespace", "dim", "vectors", "id_by_uid", "hash_by_uid"})


class Embedder(Protocol):
    """The seam: turns text into vectors. The kernel ships no concrete embedder."""

    @property
    def cache_namespace(self) -> str: ...

    def embed(self, texts: list[str]) -> list[Vector]: ...


def embed_text(node: Node) -> str:
    """The frozen per-node embedding input: title and body joined by one blank line."""
    return f"{node.title}\n\n{node.body}"


def text_hash(text: str) -> str:
    """Content-address key for the vector cache."""
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def validate_namespace(namespace: str) -> None:
    """A cache_namespace must be a safe single path segment."""
    if not namespace or namespace in (".", "..") or _NAMESPACE_RE.fullmatch(namespace) is None:
        raise ValueError(f"invalid cache_namespace {namespace!r}")


_TEXT_HASH_RE = re.compile(r"[0-9a-f]{64}")


def validate_text_hash(text_hash: str) -> None:
    """A cache key must be exactly 64 lowercase hex chars (a SHA-256 hexdigest)."""
    if _TEXT_HASH_RE.fullmatch(text_hash) is None:
        raise ValueError(f"invalid text_hash {text_hash!r}")


def _validate_finite(vec: tuple[object, ...]) -> None:
    if len(vec) < 1:
        raise ValueError("vector must have length >= 1")
    for x in vec:
        if isinstance(x, bool) or not isinstance(x, (int, float)):
            raise ValueError(f"vector contains non-finite or non-numeric value {x!r}")
        try:
            finite = math.isfinite(x)
        except OverflowError:
            finite = False
        if not finite:
            raise ValueError(f"vector contains non-finite or non-numeric value {x!r}")


def _normalize(vec: Vector) -> Vector:
    """Return the L2-normalized vector; reject zero-norm and invalid numeric input."""
    _validate_finite(vec)
    norm = math.sqrt(sum(x * x for x in vec))
    if norm == 0.0:
        raise ValueError("cannot normalize a zero-norm vector")
    return tuple(x / norm for x in vec)


@dataclass
class SimilarHit:
    id: str
    uid: str
    score: float


def _validate_k(k: int | None) -> None:
    if k is not None and (isinstance(k, bool) or not isinstance(k, int) or k <= 0):
        raise ValueError(f"k must be a positive int or None, got {k!r}")


class VectorCache:
    """Content-addressed on-disk cache of RAW embedder output, namespaced per embedder.

    Disposable: deleting the directory just forces re-embedding. All ranking math
    lives in VectorIndex; this is purely a model-output cache.
    """

    def __init__(self, root: Path | str) -> None:
        self.root = Path(root)

    def _rel(self, namespace: str, text_hash: str) -> str:
        validate_namespace(namespace)
        validate_text_hash(text_hash)
        return f"{RESERVED_NAMESPACE}/vectors/{namespace}/{text_hash}.json"

    def get(self, namespace: str, text_hash: str) -> Vector | None:
        rel = self._rel(namespace, text_hash)
        try:
            data = read_json(self.root, rel)
        except (ValueError, OSError) as exc:
            raise ValueError(f"corrupt cache file {rel}: {exc}") from exc
        if data is None:
            return None
        if not isinstance(data, dict) or "dim" not in data or "vector" not in data:
            raise ValueError(f"corrupt cache file {rel}: missing dim/vector")
        raw = data["vector"]
        if not isinstance(raw, list) or len(raw) != data["dim"]:
            raise ValueError(f"corrupt cache file {rel}: dim/vector length mismatch")
        raw_tuple = tuple(raw)
        _validate_finite(raw_tuple)
        return tuple(float(x) for x in raw_tuple)

    def put(self, namespace: str, text_hash: str, vector: Vector) -> None:
        _validate_finite(vector)
        write_json_atomic(self.root, self._rel(namespace, text_hash), {"dim": len(vector), "vector": list(vector)})


@dataclass(frozen=True)
class _PreparedVector:
    text_hash: str
    namespace: str
    vector: Vector | None  # None => content unchanged (id-only refresh)


class VectorIndex:
    """In-memory uid -> L2-normalized vector store with exact cosine ranking.

    Bound to exactly one embedder namespace and one dimension (cosine across
    vectors from different models or dimensions is meaningless).
    """

    def __init__(self) -> None:
        self.vectors: dict[str, Vector] = {}
        self.id_by_uid: dict[str, str] = {}
        self.hash_by_uid: dict[str, str] = {}
        self.dim: int | None = None
        self.namespace: str | None = None

    def to_dict(self) -> dict:
        return {
            "namespace": self.namespace,
            "dim": self.dim if self.vectors else None,
            "vectors": {uid: list(vec) for uid, vec in self.vectors.items()},
            "id_by_uid": dict(self.id_by_uid),
            "hash_by_uid": dict(self.hash_by_uid),
        }

    @classmethod
    def from_dict(cls, d: dict) -> "VectorIndex":
        idx = cls()
        if not isinstance(d, dict):
            raise ValueError("vector snapshot: document must be a dict")
        missing = _VECTOR_SNAPSHOT_KEYS - d.keys()
        if missing:
            raise ValueError(f"vector snapshot: missing {sorted(missing)[0]}")
        namespace = d["namespace"]
        vectors_raw = d["vectors"]
        id_by_uid_raw = d["id_by_uid"]
        hash_by_uid_raw = d["hash_by_uid"]
        if not isinstance(vectors_raw, dict):
            raise ValueError("vector snapshot: vectors must be a dict")
        if not isinstance(id_by_uid_raw, dict):
            raise ValueError("vector snapshot: id_by_uid must be a dict")
        if not isinstance(hash_by_uid_raw, dict):
            raise ValueError("vector snapshot: hash_by_uid must be a dict")
        if any(not isinstance(uid, str) for uid in vectors_raw):
            raise ValueError("vector snapshot: vector uid keys must be strings")
        if any(not isinstance(uid, str) or not isinstance(node_id, str) for uid, node_id in id_by_uid_raw.items()):
            raise ValueError("vector snapshot: id_by_uid must map string uids to string ids")
        for uid, text_hash_value in hash_by_uid_raw.items():
            if not isinstance(uid, str):
                raise ValueError("vector snapshot: hash_by_uid uid keys must be strings")
            if not isinstance(text_hash_value, str):
                raise ValueError("vector snapshot: hash_by_uid values must be strings")
            try:
                validate_text_hash(text_hash_value)
            except ValueError as exc:
                raise ValueError("vector snapshot: invalid hash_by_uid value") from exc
        id_by_uid = dict(id_by_uid_raw)
        hash_by_uid = dict(hash_by_uid_raw)
        if not (set(vectors_raw) == set(id_by_uid) == set(hash_by_uid)):
            raise ValueError("vector snapshot: vectors/id_by_uid/hash_by_uid uid sets differ")
        dim = d["dim"]
        if vectors_raw:
            if not isinstance(namespace, str):
                raise ValueError("vector snapshot: namespace must be a string when vectors are present")
            validate_namespace(namespace)
            if not isinstance(dim, int) or isinstance(dim, bool) or dim <= 0:
                raise ValueError("vector snapshot: dim must be a positive int when vectors are present")
        elif dim is not None:
            raise ValueError("vector snapshot: dim must be null when there are no vectors")
        elif namespace is not None:
            if not isinstance(namespace, str):
                raise ValueError("vector snapshot: namespace must be a string when non-null")
            validate_namespace(namespace)
        vectors: dict[str, Vector] = {}
        for uid, raw_vec in vectors_raw.items():
            if not isinstance(raw_vec, list):
                raise ValueError("vector snapshot: vector must be a list")
            vec_values = tuple(raw_vec)
            _validate_finite(vec_values)
            vec = tuple(float(x) for x in vec_values)
            if len(vec) != dim:
                raise ValueError("vector snapshot: vector length != dim")
            norm = math.sqrt(sum(x * x for x in vec))
            if not math.isclose(norm, 1.0, rel_tol=1e-9, abs_tol=1e-12):
                raise ValueError("vector snapshot: stored vector must be L2-normalized")
            vectors[uid] = vec
        idx.namespace = namespace
        idx.vectors = vectors
        idx.id_by_uid = id_by_uid
        idx.hash_by_uid = hash_by_uid
        idx.dim = dim
        return idx

    @classmethod
    def build(cls, nodes: Iterable[Node], embedder: Embedder, cache: VectorCache) -> "VectorIndex":
        idx = cls()
        validate_namespace(embedder.cache_namespace)
        idx.namespace = embedder.cache_namespace  # bind even for an empty corpus
        for node in nodes:
            if node.uid in idx.hash_by_uid:
                raise CollisionError(f"duplicate uid {node.uid!r} in corpus")
            idx.upsert(node, embedder, cache)
        return idx

    def prepare(self, node: Node, embedder: Embedder, cache: VectorCache) -> _PreparedVector:
        """Resolve + validate the vector WITHOUT mutating index state (cache writes ok)."""
        namespace = embedder.cache_namespace
        validate_namespace(namespace)
        if self.namespace is not None and namespace != self.namespace:
            raise ValueError(f"embedder namespace {namespace!r} != index namespace {self.namespace!r}")
        text = embed_text(node)
        h = text_hash(text)
        if self.hash_by_uid.get(node.uid) == h:
            return _PreparedVector(text_hash=h, namespace=namespace, vector=None)
        cached = cache.get(namespace, h)
        if cached is None:
            embedded = embedder.embed([text])
            if len(embedded) != 1:
                raise ValueError(f"embedder returned {len(embedded)} vectors for 1 input")
            raw_values = tuple(embedded[0])
            _validate_finite(raw_values)
            raw: Vector = tuple(float(x) for x in raw_values)
            cache.put(namespace, h, raw)
        else:
            raw = cached
        if self.dim is not None and len(raw) != self.dim:
            raise ValueError(f"vector dim {len(raw)} != index dim {self.dim}")
        return _PreparedVector(text_hash=h, namespace=namespace, vector=_normalize(raw))

    def commit(self, node: Node, prepared: _PreparedVector) -> None:
        """Apply a prepared vector. Infallible: never raises on valid prepared input."""
        if self.namespace is None:
            self.namespace = prepared.namespace
        if prepared.vector is None:
            self.id_by_uid[node.uid] = node.id  # rename / id-only refresh
            return
        if self.dim is None:
            self.dim = len(prepared.vector)
        self.vectors[node.uid] = prepared.vector
        self.id_by_uid[node.uid] = node.id
        self.hash_by_uid[node.uid] = prepared.text_hash

    def upsert(self, node: Node, embedder: Embedder, cache: VectorCache) -> None:
        self.commit(node, self.prepare(node, embedder, cache))

    def remove(self, uid: str) -> None:
        self.vectors.pop(uid, None)
        self.id_by_uid.pop(uid, None)
        self.hash_by_uid.pop(uid, None)

    def query_vector(self, vec: Vector, k: int | None = None) -> list[SimilarHit]:
        _validate_k(k)
        return self._rank(self._prepare_query(vec), k, exclude_uid=None)

    def similar(self, uid: str, k: int | None = None) -> list[SimilarHit]:
        _validate_k(k)
        if uid not in self.vectors:
            raise KeyError(uid)
        return self._rank(self.vectors[uid], k, exclude_uid=uid)

    def similar_text(self, text: str, embedder: Embedder, k: int | None = None) -> list[SimilarHit]:
        _validate_k(k)
        if self.namespace is not None and embedder.cache_namespace != self.namespace:
            raise ValueError(
                f"embedder namespace {embedder.cache_namespace!r} != index namespace {self.namespace!r}"
            )
        embedded = embedder.embed([text])
        if len(embedded) != 1:
            raise ValueError(f"embedder returned {len(embedded)} vectors for 1 input")
        return self.query_vector(tuple(embedded[0]), k)

    def _prepare_query(self, vec: Vector) -> Vector:
        raw_values = tuple(vec)
        _validate_finite(raw_values)
        if self.dim is not None and len(raw_values) != self.dim:
            raise ValueError(f"query dim {len(raw_values)} != index dim {self.dim}")
        return _normalize(tuple(float(x) for x in raw_values))

    def _rank(self, query_vec: Vector, k: int | None, *, exclude_uid: str | None) -> list[SimilarHit]:
        hits = [
            SimilarHit(
                id=self.id_by_uid[uid],
                uid=uid,
                score=sum(a * b for a, b in zip(query_vec, vec)),
            )
            for uid, vec in self.vectors.items()
            if uid != exclude_uid
        ]
        hits.sort(key=lambda h: (-score_key(h.score), h.id))
        return hits if k is None else hits[:k]
